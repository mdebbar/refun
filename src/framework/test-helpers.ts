import { CommitNode, Commit } from './core';

let incrementalCommitValue = 0;

// This function has to be called at the beginning of any test that uses
// jest snapshots with commits.
export function resetCommitValue() {
  incrementalCommitValue = 0;
}

export class TestCommit<N extends CommitNode> extends Commit<N> {
  // Can be used to check the order of commit creation and application.
  $started: number;
  $applied: number;

  constructor(node: N) {
    super(node);
    this.$started = incrementalCommitValue++;
  }

  didApply() {
    this.$applied = incrementalCommitValue++;
  }

  addChild(commit: Commit): boolean {
    return super.addChild(commit);
  }
}

export class MockCommitNode extends CommitNode {
  startCommit(): TestCommit<this> {
    return new TestCommit(this);
  }

  applyCommit(commit: TestCommit<this>) {
    commit.didApply();
  }
}

// export class TrackedCommitNode extends CommitNode {
//   appliedCommits: Commit<this>[];

//   constructor() {
//     super();
//     this.appliedCommits = [];
//   }

//   startCommit(): TestCommit<this> {
//     return new TestCommit(this);
//   }
//   applyCommit(commit: TestCommit<this>) {
//     this.appliedCommits.push(commit);
//     commit.didApply();
//   }
// }

export class RootTestNode extends MockCommitNode {}
