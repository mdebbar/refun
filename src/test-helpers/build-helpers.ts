import { AppNode, myStateNode, UI, namedComponent } from '../framework';

export function createRef(name: string) {
  return namedComponent(name, (children: () => UI) => {
    myStateNode(null);
    return children();
  });
}

export function manageFrames(rootNode: AppNode<any>): VoidFunction {
  let _cb: VoidFunction | null = null;
  rootNode.scheduleNextFrame = cb => {
    _cb = cb;
  };

  return function nextFrame() {
    if (_cb == null) {
      throw new Error('There was no scheduled rebuild');
    }
    _cb();
    _cb = null;
  };
}
