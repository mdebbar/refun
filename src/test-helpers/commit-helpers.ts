import { Committer, myCommitNode, component, UI } from '../framework';

export class TestCommit {
  // Can be used to check the order of commit creation and application.
  value: string;
  children: TestCommit[];

  constructor(value: string) {
    this.value = value;
    this.children = [];
  }
}

export class TestCommitter extends Committer<TestCommit> {
  private data: string;
  private testCommit: TestCommit | null = null;

  constructor(data: string) {
    super();
    this.data = data;
  }

  getTestCommit(): TestCommit {
    return this.testCommit as TestCommit;
  }

  update(newData: string) {
    this.data = newData;
    this.needsCommit();
  }

  commitSelf(lastCommit: TestCommit, newChildren: TestCommit[]): TestCommit {
    if (!lastCommit) {
      return new TestCommit(this.data);
    }
    lastCommit.value = `${lastCommit.value}|${this.data}`;
    return lastCommit;
  }

  diffChildren(newCommit: TestCommit, newChildren: TestCommit[]): void {
    newCommit.value = newCommit.value + '*';
    this.testCommit = newCommit;
    newCommit.children = newChildren;
  }
}

export const committer = component(function committer(
  str: string,
  children: UI,
) {
  const node = myCommitNode(() => new TestCommitter(str));
  node.committer.update(str);
  return children;
});
