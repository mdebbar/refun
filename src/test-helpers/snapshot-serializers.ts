import { AppNode, CommitNode } from '../framework';
import { TestCommitter, TestCommit } from './commit-helpers';

type any2s = (value: any) => string;
type s2s = (str: string) => string;

type Children = any[];

function serializeChildren(
  children: Children,
  serialize: any2s,
  indent: s2s,
): string {
  if (children.length === 0) {
    return '[]';
  }
  return ['[', indent(children.map(serialize).join(',\n')), ']'].join('\n');
}

export const stateSerializer = {
  test(value: any) {
    return value instanceof AppNode;
  },

  print(value: AppNode<any>, serialize: any2s, indent: s2s): string {
    const committer =
      value instanceof CommitNode && value.committer instanceof TestCommitter
        ? value.committer
        : null;
    return [
      `%${value.a__name}%`,
      serialize(value.state),
      ...(committer
        ? [`commit value: ${committer.getTestCommit().value}`]
        : []),
      serializeChildren(value.children, serialize, indent),
    ].join('\n');
  },
};

export const commitSerializer = {
  test(value: any) {
    return value instanceof TestCommit;
  },

  print(value: TestCommit, serialize: any2s, indent: s2s): string {
    return (
      `#${value.value} ` + serializeChildren(value.children, serialize, indent)
    );
  },
};
