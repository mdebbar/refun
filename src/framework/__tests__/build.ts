import { run, stateful, AppNode, UI, findDescendantByName } from '../';
import { RootTestNode, resetCommitValue } from '../test-helpers';

const app = (children: UI) => container(children);
const container = stateful('tester', (_: AppNode, children: UI) => children);

test('simple build tree', () => {
  const root = run(app(null), new RootTestNode());
  expect(root).toMatchSnapshot();
});

test('nested arrays', () => {
  resetCommitValue();
  const ui1 = app(null);
  expect(run(ui1, new RootTestNode())).toMatchSnapshot();

  resetCommitValue();
  const ui2 = app([[null]]);
  // This snapshot should be identical to the one above.
  expect(run(ui2, new RootTestNode())).toMatchSnapshot();
});

test('complex arrays', () => {
  resetCommitValue();
  const dup = (children: UI) => [children, children];
  const ui = dup(dup([null, dup(null)]));
  const root = run(ui, new RootTestNode());
  expect(root).toMatchSnapshot();
});

test('preserves state when rebuilding', () => {
  resetCommitValue();
  const counter = stateful('counter', (node: AppNode) => {
    if (node.state == null) {
      node.state = 0;
    }
    node.state++;
    return doubleCounter();
  });

  const doubleCounter = stateful('doubleCounter', (node: AppNode) => {
    if (node.state == null) {
      node.state = 0;
    }
    node.state += 2;
    return null;
  });

  const root = run(app(counter()), new RootTestNode());

  const counterNode = findDescendantByName(root, 'counter');
  const doubleCounterNode = findDescendantByName(root, 'doubleCounter');

  expect(counterNode.state).toBe(1);
  expect(doubleCounterNode.state).toBe(2);

  // Rebuild from the root.
  root.rebuild();
  // App nodes should be preserved.
  expect(findDescendantByName(root, 'counter')).toBe(counterNode);
  expect(findDescendantByName(root, 'doubleCounter')).toBe(doubleCounterNode);
  // States should've updated.
  expect(counterNode.state).toBe(2);
  expect(doubleCounterNode.state).toBe(4);

  // Rebuild from the counter node.
  counterNode.rebuild();
  expect(counterNode.state).toBe(3);
  expect(doubleCounterNode.state).toBe(6);

  // Rebuild from the double counter node.
  doubleCounterNode.rebuild();
  expect(counterNode.state).toBe(3);
  expect(doubleCounterNode.state).toBe(8);
});

test('complex state', () => {
  resetCommitValue();
  const counter = stateful(
    'counter',
    (node: AppNode, cb: (x: number) => UI) => {
      if (node.state == null) {
        node.state = 0;
      }
      node.state++;
      return cb(node.state);
    },
  );

  const doubleCount = stateful('doubleCount', (_: AppNode, count: number) => {
    _.state = count * 2;
    return null;
  });

  const root = run(app(counter(doubleCount)), new RootTestNode());

  const counterNode = findDescendantByName(root, 'counter');
  const doubleCountNode = findDescendantByName(root, 'doubleCount');

  expect(counterNode.state).toBe(1);
  expect(doubleCountNode.state).toBe(2);

  // Rebuild from the root.
  root.rebuild();
  expect(counterNode.state).toBe(2);
  expect(doubleCountNode.state).toBe(4);

  // Rebuild from the doubleCount node should keep the same state.
  doubleCountNode.rebuild();
  expect(counterNode.state).toBe(2);
  expect(doubleCountNode.state).toBe(4);
});
