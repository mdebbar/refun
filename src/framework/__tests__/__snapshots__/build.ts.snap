// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex arrays 1`] = `
AppNode {
  "children": Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "commitNode": RootTestNode {
    "lastCommit": TestCommit {
      "$applied": 1,
      "$started": 0,
      "children": Array [],
      "node": [Circular],
    },
  },
  "parent": null,
  "render": [Function],
  "state": null,
}
`;

exports[`nested arrays 1`] = `
AppNode {
  "children": Array [
    AppNode {
      "children": Array [
        null,
      ],
      "commitNode": null,
      "parent": [Circular],
      "render": [Function],
      "state": null,
    },
  ],
  "commitNode": RootTestNode {
    "lastCommit": TestCommit {
      "$applied": 1,
      "$started": 0,
      "children": Array [],
      "node": [Circular],
    },
  },
  "parent": null,
  "render": [Function],
  "state": null,
}
`;

exports[`nested arrays 2`] = `
AppNode {
  "children": Array [
    AppNode {
      "children": Array [
        null,
      ],
      "commitNode": null,
      "parent": [Circular],
      "render": [Function],
      "state": null,
    },
  ],
  "commitNode": RootTestNode {
    "lastCommit": TestCommit {
      "$applied": 1,
      "$started": 0,
      "children": Array [],
      "node": [Circular],
    },
  },
  "parent": null,
  "render": [Function],
  "state": null,
}
`;

exports[`simple build tree 1`] = `
AppNode {
  "children": Array [
    AppNode {
      "children": Array [
        null,
      ],
      "commitNode": null,
      "parent": [Circular],
      "render": [Function],
      "state": null,
    },
  ],
  "commitNode": RootTestNode {
    "lastCommit": TestCommit {
      "$applied": 1,
      "$started": 0,
      "children": Array [],
      "node": [Circular],
    },
  },
  "parent": null,
  "render": [Function],
  "state": null,
}
`;
